src/providers/ipa/ipa_subdomains_ext_groups.c

```diff
@@ -326,62 +334,169 @@ static errno_t add_ad_user_to_cached_groups(struct ldb_dn *user_dn,
+    }
+
+    /* TODO: there is a similar functionality (adding and removing group
+     * memberships in sysdb_update_members_ex(), but the missing group feature
+     * is missing. It might be worth to evaluate if either the missing group
+     * feature can be added there or if group which are missing in the cache
+     * can bew handled differently here. */
+
+    for (c = 0; add_groups[c] != NULL; c++) {
+
+        ret = sysdb_search_groups_by_orig_dn(tmp_ctx, group_dom, add_groups[c],
                                              NULL, &msgs_count, &msgs);
         if (ret != EOK) {
             if (ret == ENOENT) {
                 DEBUG(SSSDBG_TRACE_ALL, "Group [%s] not in the cache.\n",
-                                         groups[c]);
+                                         add_groups[c]);
                 *missing_groups = true;
+                talloc_free(groups[d]);
+                /* add missing group back to the list */
+                groups[d++] = talloc_steal(groups, add_groups[c]);
                 continue;
             } else {
                 DEBUG(SSSDBG_OP_FAILURE, "sysdb_search_entry failed.\n");
                 goto done;
             }
         }
 
-/* TODO? Do we have to remove members as well? I think not because the AD
- * query before removes all memberships. */
-
         ret = sysdb_mod_group_member(group_dom, user_dn, msgs[0]->dn,
                                      LDB_FLAG_MOD_ADD);
         if (ret != EOK && ret != EEXIST) {
             DEBUG(SSSDBG_OP_FAILURE, "sysdb_mod_group_member failed.\n");
             goto done;
         }
 
-        user_attrs = sysdb_new_attrs(tmp_ctx);
-        if (user_attrs == NULL) {
-            DEBUG(SSSDBG_OP_FAILURE, "sysdb_new_attrs failed.\n");
-            ret = ENOMEM;
+    }
+    talloc_free(groups[d]);
+    groups[d] = NULL;
+
+    for (c = 0; del_groups[c] != NULL; c++) {
+        ret = sysdb_search_groups_by_orig_dn(tmp_ctx, group_dom, del_groups[c],
+                                             NULL, &msgs_count, &msgs);
+        if (ret != EOK) {
+            if (ret == ENOENT) {
+                DEBUG(SSSDBG_TRACE_ALL,
+                      "Group [%s] not in the cache, skipping.\n",
+                      del_groups[c]);
+                continue;
+            } else {
+                DEBUG(SSSDBG_OP_FAILURE, "sysdb_search_entry failed.\n");
+                goto done;
+            }
+        }
+
+        ret = sysdb_mod_group_member(group_dom, user_dn, msgs[0]->dn,
+                                     LDB_FLAG_MOD_DELETE);
+        if (ret != EOK && ret != EEXIST) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "sysdb_mod_group_member failed to delete member.\n");
             goto done;
         }
 
         ret = sysdb_attrs_add_string(user_attrs, SYSDB_ORIG_MEMBEROF,
-                                     groups[c]);
+                                     del_groups[c]);
```

Comment:
I'm lost here: why do we add deleted group membership?

Reply:
`removed` is fixed. The `sysdb_attrs_add_string()` is a left-over I forgot to remove, `SYSDB_ORIG_MEMBEROF` is added before in this function.

Reply:
SYSDB_ORIG_MEMBEROF is added based on 'groups' to reflect actual membership.
Here it was being additionally done using 'del_groups', so was puzzled if I totally misunderstand the patch.
Thanks for the update.